Rework LayeredMaterial

gchoqueux opened the pull request:
- **Add `Style` object to WFS**
- **Rework LayeredMaterial**
  - All textures and parameters are ordered by display order.
  - `this.Layers` contains all layers by display order
  - The slots's texture management is done internally, now, to set texture uses `layerId`
  - Fix bug : Remove color layer doesn't dispose textures colors
- **Uniform changes** (`LayeredMaterial` and `BasicMaterial`)
  - use`THREE.Uniform` in material
  - `int RTC`          becomes `bool useRTC`
  - `int lightingOn` becomes `bool lightingOn`
  - `int debug`       becomes `bool debug`
  - `int lightingOn` becomes `bool lightingOn`
  - `int selected`    becomes `bool selected`
- **Remove slot texture** in `NodeProcess` is replaced by `layerId`
- **Change Skirt size** influence performance rendering, minimize pixel shader pass
- **Fix typos and add Binary Tools**

gchoqueux committed:
Change Skirt size -> influence performance

gchoqueux committed:
Add StyleManager object
Add Style object
Add Palette object

peppsac commented on src/Globe/Globe.js:
```diff
@@ -208,6 +208,17 @@ Globe.prototype.updateLayersOrdering = function() {
+Globe.prototype.removeColorLayer = function(layer) {
+
+    var cO = function(object) {
+        if (object.changeSequenceLayers)
```
Add missing `{  }`

peppsac replied:
Didn't you intend to write `if (object.removeColorLayer)` ?

peppsac commented on src/Globe/TileMesh.js:
```diff
@@ -320,42 +281,25 @@ TileMesh.prototype.loadingCheck = function() {
+TileMesh.prototype.removeColorLayer = function(idLayer) {
+    var texturesCount = this.materials[RendererConstant.FINAL].loadedTexturesCount();
+    this.materials[RendererConstant.FINAL].removeColorLayer(idLayer);
+    this.texturesNeeded -= texturesCount - this.materials[RendererConstant.FINAL].loadedTexturesCount();
```
This is wrong.
If the textures for this layer aren't yet loaded  this tile is going to be stuck in the `!loaded` state because `this.texturesNeeded` will be incorrect.

`texturesNeeded` should be updated the same way it's initialized in `NodeProcess`.
A simpler alternative would be to disallow layer removal for `!loaded` tiles.

gchoqueux replied:
I rework this part

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -177,152 +187,248 @@ LayeredMaterial.prototype.dispose = function() {
 
-LayeredMaterial.prototype.setNbLayersColor = function(n) {
-    this.uniforms.nColorLayer.value = n;
+LayeredMaterial.prototype.downScaledColorLayer = function(layer,level) {
```
Maybe rename `isColorLayerDownscaled`?

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -177,152 +187,248 @@ LayeredMaterial.prototype.dispose = function() {
+    return this.Textures[l_COLOR][this.getLayerTextureOffset(layer)].level < level;
+};
+
+LayeredMaterial.prototype.downScaledLayer = function(layerType,level) {
```
Maybe rename `isLayerTypeDownscaled`?

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -177,152 +187,248 @@ LayeredMaterial.prototype.dispose = function() {
-LayeredMaterial.prototype.getLayerLevel = function(id, slot) {
-    var level = this.Textures[id][slot || 0].level;
+    let index = this.indexOfColorLayer(colorLayerId);
+    if(index ===-1) {
```
I'm not sure that hiding the "colorLayerId is invalid" error is a good idea.
Either throw an error, or return an invalid level value (`-1` ?)

gchoqueux replied:
I prefer work this in another PR

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -95,36 +95,43 @@ NodeProcess.prototype.subdivideNode = function(node, camera, params) {
 
                 child.matrixSet = [];
-
+                let tileMatrixSet,texturesCount;
```
Nope. Replacing `var` by `let` is a good idea but don't move the declaration to a bigger scope.

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -177,152 +187,248 @@ LayeredMaterial.prototype.dispose = function() {
+
+    this.uniforms.paramLayers.value[newIndex] = new THREE.Vector4();
+
+    this.setLayerOffset(newIndex,offset);
```
You could rename `setLayerOffset` : `setTextureOffsetByLayerIndex` to match its getter

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -176,16 +183,17 @@ NodeProcess.prototype.refineNodeLayers = function(node, camera, params) {
-        if (node.pendingLayers[i] === undefined
-            && (!node.loaded || node.downScaledLayer(i))) {
+    // WARNING don't use declaration function to reset flag
+    // var resetPendingFlag = function(id){node.pendingLayers[id] = undefined;};
```
Remove this

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -176,16 +183,17 @@ NodeProcess.prototype.refineNodeLayers = function(node, camera, params) {
+    // var resetPendingFlag = function(id){node.pendingLayers[id] = undefined;};
 
-            node.pendingLayers[i] = true;
+    for (let typeLayer=0; typeLayer<2; typeLayer++) {
```
It's probably a good call to rename `i` -> `layerType` but it could probably be defered to a specific cleanup PR.

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -205,18 +213,12 @@ NodeProcess.prototype.hideNodeChildren = function(node) {
-            return null;
+    if (parent && parent.material && parent.material.getLayerLevel) {
+        var level = parent.material.getLayerLevel(layerType, layer);
+        if (level >= 0) {
```
I'd prefer to keep the other order `0 <= level`
I like to only use `<` and `<=` and never use `>`. Of course in single statement it doesn't matter, but in more complex case :

``` js
if (a >= -2 && a <= 10)
if (-2 <= a && a <= 10)
```

I think the 2nd version is easier to read.

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -177,152 +187,248 @@ LayeredMaterial.prototype.dispose = function() {
+    return false;
 };
 
+LayeredMaterial.prototype.getColorLayerLevel = function(colorLayerId) {
```
`getColorLevelByLayerId` ?
`getColorLayerLevelById` ?

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,132 +43,140 @@ var getColorAtIdUv = function(nbTex) {
-    this.paramLayers = Array(8);
-    this.paramBLayers = Array(8);
+    var paramLayers = Array(8);
+    this.layerTexturesCount = Array(8);
```
This is `colorLayerTexturesCount`.
Also it's a bit annoying to duplicate texture count in `layerTextureCount` and `texturesCount` :(

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -324,18 +322,9 @@ function updateNodeElevation(quadtree, node, layersConfig, force) {
-                    break;
-                }
-            }
+        let a = findAncestorWithValidTextureForLayer(node, l_ELEVATION);
```
`ancestor = findAncestorWithValidTextureForLayer(node, l_ELEVATION);`

peppsac replied:
Hmm you don't set `bestLayer`, is this on purpose?

gchoqueux replied:
`bestLayer` is found in loop [here](https://github.com/iTowns/itowns2/blob/6826b6c61d485dc7f9023ca3a282111c81da9c88/src/Scene/NodeProcess.js#L353)
Maybe rework this in other PR with `LayeredMaterial.prototype.getColorLayerLevelById`

peppsac replied:
bestLayer was supposed to be set at the same time than the ancestor. Or the 1st pass thing is useless (it's supposed to retrieve ancestor inital texture)

gchoqueux replied:
In `LayeredMaterial`, for the moment _elevation layer_ haven't the same mechanism that _color layer_ . _Elevation layer_ isn't generic like _color layer_. It's not possible for the moment to find `bestLayer` with `ancestor`. Because the TileMesh, LayeredMaterial or texture haven't information on  _elevation layer_'s identifiant. 
We need a new PR to make _Elevation layer_ more generic and and to superimpose _elevation layers_.

peppsac replied:
Ok I get your point.
I agree that _elevation_ handling needs to be reworked in different PR (maybe issue #192 is relevant btw).

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -344,31 +333,29 @@ function updateNodeElevation(quadtree, node, layersConfig, force) {
-            if (!layer.tileInsideLimit(node, layer)) {
-                continue;
-            }
+    for (let i = elevationLayers.length - 1; i >= 0; i--) {
```
I don't really mind changing the order here, but if it's important add a comment stating why we must browse elevation layer in this order.

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -415,24 +402,19 @@ NodeProcess.prototype.processNode = function(node, camera, params) {
-        }
-
-        if (!hidden)  {
+        } else if (!hidden && node.downScaledLayer(1)) {
```
Don't use `1` magic value.
Also, are you not interested in refining elevation layers?

gchoqueux commented:
Updated and rebased

gchoqueux committed:
Rework LayeredMaterial and uniforms shaders
- Rework LayeredMaterial
  * All textures and parameters are ordered by display order.
  * `this.Layers` contains all layers by display order
  * The slots's texture management is done internally, now, to set texture uses `layerId`
  * Fix bug : Remove color layer doesn't dispose textures colors

- Uniform changes (`LayeredMaterial` and `BasicMaterial`)
  * use`THREE.Uniform` in material
  * `int RTC` becomes `bool useRTC`
  * `int lightingOn` becomes `bool lightingOn`
  * `int debug` becomes `bool debug`
  * `int lightingOn` becomes `bool lightingOn`
  * `int selected` becomes `bool selected`

gchoqueux committed:
Add binary tools object
Fix error parameter in WMTS
Remove warnings and fix typos
index.html clean up

gchoqueux commented:
Rebased and fix problem ancestor elevation

gchoqueux commented:
updated and rebased

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -324,51 +318,41 @@ function updateNodeElevation(quadtree, node, layersConfig, force) {
+            continue;
+        }
 
+        if(ancestor === null) {
```
This change is weird and the behavior is not obvious.
I understand why you did this but this is acceptable only because it's meant to be temporary.

Add a comment explaining what is going on (either here or when you define `ancestor` line 321):

``` js
// ancestor is not enough: we also need to know from which layer we're going to request the elevation texture (see how this is done for color texture).
// Right now this is done in the `for` loop below but this is hacky because there's no real warranty that bestLayer and ancestor really match.
// FIXME: we need to be able to set both ancestor and bestLayer at the same time
```

peppsac commented on src/Globe/Globe.js:
```diff
@@ -208,6 +208,17 @@ Globe.prototype.updateLayersOrdering = function() {
+        if (object.removeColorLayer) {
+            object.removeColorLayer(layer);
+        }
+    }.bind(this);
```
Is `bind(this)` needed?

peppsac commented on src/Globe/TileMesh.js:
```diff
@@ -308,54 +269,38 @@ TileMesh.prototype.OBB = function() {
 
-    var sequenceMaterial = [];
+    // Quit if there is only one layer
+    if (layerCount < 2)
```
Missing `{ }`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,288 +43,400 @@ var getColorAtIdUv = function(nbTex) {
-        this.paramLayers[j].x -= nbTextures;
+    // Dispose Layers textures
+    for (i = offset, max = offset + texturesCount; i < max; i++) {
+        if (this.textures[l_COLOR][i] instanceof THREE.Texture)
```
Missing `{ }`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,288 +43,400 @@ var getColorAtIdUv = function(nbTex) {
-        if (this.Textures[1][i] instanceof THREE.Texture)
-            this.Textures[1][i].dispose();
+    if (layerIndex === -1) {
+        return;
```
At some point we'll need to stop silently ignoring errors and decide on a strategy (log errors ? throw ?)

gchoqueux replied:
It's necessary to choose a solution

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,288 +43,400 @@ var getColorAtIdUv = function(nbTex) {
-    this.layerIdToIndex.idLayer = undefined;
-};
+    // Update slot start texture layer
+    for (var j = layerIndex, mx = this.getColorLayersCount(); j < mx; j++)
```
Missing `{ }`

peppsac replied:
Also use `let i` instead of `var j`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,288 +43,400 @@ var getColorAtIdUv = function(nbTex) {
+LayeredMaterial.prototype.setTexturesLayer = function(textures, layerType, layer) {
+    var index = this.indexOfColorLayer(layer);
+    var slotOffset = this.getTextureOffsetByLayerIndex(index);
+    for (var i = 0, max = textures.length; i < max; i++) {
```
Always use `let` for new code

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,288 +43,400 @@ var getColorAtIdUv = function(nbTex) {
-            return l-1;
+LayeredMaterial.prototype.setColorLayerParameters = function(params) {
+    if(this.getColorLayersCount() === 0) {
+        for (var l = 0; l < params.length; l++) {
```
style: `let` and `if (` (missing space)

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,288 +43,400 @@ var getColorAtIdUv = function(nbTex) {
+            return this.textures[l_ELEVATION][0].level < level;
+        }
+    } else if (layerType === l_COLOR) {
+        for (var index = 0, max = this.colorLayersId.length; index < max; index++) {
```
`let` :)

gchoqueux commented:
updated and rebased

peppsac commented on src/Globe/TileGeometry.js:
```diff
@@ -213,15 +214,16 @@ TileGeometry.prototype.computeBuffers = function(params, builder) {
+
+    //TODO: WARNING beware size'skirt influence performance
+    var rmax = 50;
+    var r = Math.pow(rmax, 1 + 1.25/ params.zoom);
```
Why did you change the formula to `1+1.25/zoom`?

gchoqueux replied:
I simplify the formula. 
Skirt's size influence performance. For the moment, I reduce the size to increase performance (pixel shader performance)
But this calculation isn't correct. A correct calculation must to minimize the size without crack between tiles. This size must be take into account the bbox's size.

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,288 +43,402 @@ var getColorAtIdUv = function(nbTex) {
 
-    this.Textures[1].splice(startIdTexture, nbTextures);
-    this.uniforms.nColorLayer.value--;
+    var offset = this.getTextureOffsetByLayerIndex(layerIndex);
```
`var` -> `let`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,288 +43,402 @@ var getColorAtIdUv = function(nbTex) {
-    // Rebuild sequence
-    var sequence = this.uniforms.layerSequence.value;
-    var limit = false;
+    var removedTexturesLayer = this.textures[l_COLOR].splice(offset, texturesCount);
```
`var` -> `let`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,288 +43,402 @@ var getColorAtIdUv = function(nbTex) {
+        if(-1 < removedTexturesLayer[i].level) {
+            loadedTexturesLayerCount++;
         }
+    }
```
For loops like this one can be shortened using `reduce` and arrow functions.

Something like this (not tested):

``` js
let loadedTexturesLayerCount = removedTexturesLayer.reduce((sum, texture) => sum + (-1 < texture.level), 0);
```

gchoqueux replied:
ok, I change this

peppsac commented:
I left a few comments regarding style.
But I have a more important one: I think this PR introduces a regression in texture loading/flickering.

Here's an example of the issue. During zooming you can see the time where the child tile used a worse quality texture than its parent (red square):

![crappy gif showing the bug](https://cloud.githubusercontent.com/assets/2198295/20406551/0d172112-ad0e-11e6-8744-d7ae1fd500d4.gif)

gchoqueux commented:
> I think this PR introduces a regression in texture loading/flickering.

Seems fixed

gchoqueux commented:
updated and rebased

peppsac commented:
LGTM, merging.
